keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1606 @@
+resources would be communicating the Kubernetes Pod QoS class from
+kubelet to the container runtime.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Make it possible to request QoS-class resources
+  - Support RDT class assignment of containers. This is already supported by
+    the containerd and CRI-O runtime and part of the OCI runtime-spec
+  - Support blockio class assignment of containers.
+  - Support Pod-level (sandbox-level) QoS-class resources
+- Make the API to support updating QoS-class resource assignment of running containers
+- Make the extensions flexible, enabling simple addition of other QoS-class
+  resource types in the future.
+- Make QoS-class resources opqaue (as possible) to the CRI client
+- Discovery of the available QoS-class resources
+- API changes to support updating Pod-level (sandbox-level) QoS-class resource
+  assignment of running pods ([future work](#future-work))
+- Resource status/capacity ([future work](#future-work))
+- Access control ([future work](#future-work))
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Interface or mechanism for configuring the QoS-class resources (responsibility of
+  the container runtime).
+- Enumerating possible (QoS-class) resource types or their detailed behavior
+
+## Implementation phases
+
+This proposal splits the full implementation of QoS-class resources into
+multiple phases, building functionality gradually, step-by-step. The goal is to
+make the discussions more focused and easier. We may also learn on the way,
+insights from earlier phases affecting design choises made in the later phases,
+hopefully resulting in a better overall end result. However, we also outline
+all the future steps to not lose the overall big picture.
+
+The figure below illustrates the design of the full implementation (less quota)
+and the division of implementation phases. The first implementation phase
+basically covers the communication between kubelet and the container runtime
+(i.e. CRI API). All changes to the Kubernetes API and its control plane
+components are left to future work. This KEP (the [Proposal](#proposal) and
+[Design Details](#design-details)) in its current form implements this first
+phase – the KEP will evolve and be supplemented with future phases getting
+implemented.
+
+![design](./design.svg)
+
+In the current design QoS-class resources are designed to be opaque to the CRI
+client in the sense that the container runtime takes care of configuration and
+control of the resources and the classes within.
+
+### Phase 1
+
+The goal is to enable a bare minimum for users to leverage QoS-class resources
+and start experimenting with them in Kubernetes:
+
+- extend the CRI protocol to allow QoS-class resource assignment and updates to
```

Comment:
I am not sure why we need the CRI changes for phase 1. It can be done just with annotations.

Reply:
Thanks @mrunalp. Do you mean 
1. Use pod annotations that would be interpreted by the runtime(s). I.e. we would specify annotations in `api/core/v1/` and those would be entirely handled by the runtime.
2. **OR** Use container annotations instead of CRI. I.e. pod annotations would be interpreted by kubelet and "translated" to corresponding container annotations.

I think the CRI changes are not mandatory, we can have some level of support with annotations alone. Advantages of bringing in CRI changes would be e.g.
- better "UX"
  - better error messages as kubelet can do input validation, e.g.. reject pod if requested QoS-class resource is not available
  - wrt option 1 above, being able to reject changes to the annotations of running pods (keep annotations in sync with the actual assignment)
- CRI could be used to communicate the pod qos class explicitly to the runtime (this is currently called out in open questions section). Could be used to eliminate quessing of the pod qos class e.g. in [cri-o](https://github.com/cri-o/cri-o/blob/edf889bd277ae9a8aa699c354f12baaef3d9b71d/internal/runtimehandlerhooks/high_performance_hooks.go#L223)
- prepare runtimes early on ("full" support would be available when K8s api changes are landing)

Reply:
@mrunalp I'm resolving this comment as outdated as this aspect of the proposal has completely changed (no annotations anymore)